File: packages/python/src/ghagen/lint/rules/missing_permissions.py
"""Rule: flag workflows that don't explicitly declare permissions.

GitHub's default ``GITHUB_TOKEN`` has broad write access. Setting an
explicit ``permissions:`` block at the workflow (or per-job) level
follows OWASP's supply-chain hardening guidance.
"""

from __future__ import annotations

from collections.abc import Iterable

from ghagen.lint.rules._base import RuleContext, rule
from ghagen.lint.violation import Severity, Violation
from ghagen.models.workflow import Workflow


@rule(
    id="missing-permissions",
    severity=Severity.WARNING,
    description=(
        "Workflow has no explicit permissions set. The default GITHUB_TOKEN "
        "has broad write access; an explicit permissions block limits scope."
    ),
)
def check_missing_permissions(wf: Workflow, ctx: RuleContext) -> Iterable[Violation]:
    """Flag workflows without top-level permissions unless every job sets them."""
    if wf.permissions is not None:
        return

    # Check if every job sets its own permissions — in that case,
    # the workflow-level omission is fine.
    if wf.jobs and all(
        getattr(job, "permissions", None) is not None for job in wf.jobs.values()
    ):
        return

    severity = ctx.config.severity.get(
        "missing-permissions", check_missing_permissions.meta.default_severity
    )
    yield Violation(
        rule_id="missing-permissions",
        severity=severity,
        message=(
            f"Workflow '{wf.name or ctx.workflow_key}' has no "
            "top-level permissions set."
        ),
        location=ctx.loc(wf, f"{ctx.workflow_key}.yml"),
        hint=(
            'Add permissions=Permissions(contents="read") (or similar) to '
            "limit the default GITHUB_TOKEN scope, or set permissions on "
            "every job individually."
        ),
    )

File: packages/typescript/src/lint/rules/missing-permissions.ts
/**
 * Rule: flag workflows that don't explicitly declare permissions.
 *
 * GitHub's default GITHUB_TOKEN has broad write access. Setting an
 * explicit `permissions:` block at the workflow (or per-job) level
 * follows OWASP's supply-chain hardening guidance.
 */

import { isModel, type Model } from "../../models/_base.js";
import type { Rule } from "./_base.js";
import { rule } from "./_base.js";

const META = {
  id: "missing-permissions",
  defaultSeverity: "warning" as const,
  description:
    "Workflow has no explicit permissions set. The default GITHUB_TOKEN " +
    "has broad write access; an explicit permissions block limits scope.",
};

export const checkMissingPermissions: Rule = rule(META, function* (wf, ctx) {
  if (wf._data["permissions"] !== undefined) return;

  const jobs = (wf._data["jobs"] ?? {}) as Record<string, unknown>;
  const jobValues = Object.values(jobs).filter(isModel) as Model[];

  // If every job has its own permissions, the workflow-level omission is fine.
  if (jobValues.length > 0 && jobValues.every((j) => j._data["permissions"] !== undefined)) {
    return;
  }

  const severity = ctx.config.severity.get(META.id) ?? META.defaultSeverity;
  const name = (wf._data["name"] as string | undefined) ?? ctx.workflowKey;
  yield {
    ruleId: META.id,
    severity,
    message: `Workflow '${name}' has no top-level permissions set.`,
    location: ctx.loc(wf, `${ctx.workflowKey}.yml`),
    hint:
      "Add permissions: { contents: 'read' } (or similar) to limit the " +
      "default GITHUB_TOKEN scope, or set permissions on every job " +
      "individually.",
  };
});
